compiler/rustc_parse/src/parser/mod.rs

```diff
@@ -1578,3 +1588,10 @@ pub enum NtOrTt {
     Nt(Nonterminal),
     Tt(TokenTree),
 }
+
+enum ItemOrExprBuiltin<'a> {
+    // For now, there aren't any builtin#foo constructs expanding to items.
+    #[allow(dead_code)]
+    Item(fn(&mut Parser<'a>, Span) -> PResult<'a, item::ItemInfo>),
+    Expr(fn(&mut Parser<'a>, Span) -> PResult<'a, P<Expr>>),
+}
```

Comment:
This is sort of looks like creating `ItemOrTypeImpl` for `impl Foo for Bar {}` items and `imp Trait` types.

There's nothing common between `builtin # offset_of` and `builtin # global_asm`, we should probably do the +1/+2 lookahead and disambiguate right away.
Without backward and forward compatibility concerns `offset_of` and `global_asm` would just be two keywords (`builtin` in our case only becomes a keyword so these two don't have to), and the parser logic can be structured accordingly.

Reply:
> There's nothing common between `builtin # offset_of` and `builtin # global_asm`

My reasoning for the code being this weird is that I want to make nicer errors. Correct me if my reasoning is wrong in any point:

If someone puts `builtin # some_future_builtin_syntax`, I want to make "this builtin# syntax is not known"-ish errors, instead of giving "# was unexpected character"-ish errors. Before trying to parse something as expression, the parser tries to parse it as an item. Therefore, the item parser can't yield errors for every builtin syntax that's not creating an item, it can only do that for builtin syntax that's neither yielding an expression nor an item (and it has to yield errors at all because if it doesn't then in item-only contexts, say outside of a function, putting an invalid `builtin #` syntax won't give a nice error).

Therefore, the item parser has to be able to distinguish between some `builtin #` syntax being unknown, and some `builtin #` syntax being an expression.

Trying to explain this made me aware of a flaw in my scheme though: what happens when someone tries to put a known `builtin # offset_of` for example in a context that doesn't allow expressions... do you know of a way to check for whether expressions are allowed right now?

Reply:
The alternative that I can think of would be to have different introductory syntaxes for builtins that expand to items vs ones that expand to expressions. like `builtin_item # global_asm` vs `builtin # asm` or such. I think a unified approach is nicer but shrug if you prefer that one, tell me.

Reply:
>make nicer errors

I think this is the point where the reasoning is wrong.
This syntax will almost never be used by people, so nicer errors are not needed.